### PR TITLE
JMH SocketIO Benchmark: Plain vs ZIO vs ZArrow

### PR DESCRIPTION
**Pull request by tampler**
Here I compare a runtime performance for the real world load testing scenario, which includes both effectful IO and compute intensive operations. 

The main execution unit is a plain `worker` method, which does the following:
1. Accepts temp file name 
2. Reads this file (performs effectful IO) to get a seed 
3. Calculates a simple factorial function for that seed
4. Returns a calculated factorial value

Scenario 1: Plain compute 
```scala
   val workers: List[Int]  = files.map(f => worker(f._1))
   val sumPlain: Int = workers.sum
```
Scenario 2: ZIO Effect
```scala
  val sumMon: ZIO[Any, Throwable, Int] = for {
    list <- ZIO.traverse(files) { item =>
               ZIO.effect(worker(item._1))
              }
    out = list.sum
    } yield out
```
Scenario 3: ZIO Arrow 
```scala
   val arrWorkers: ZArrow[Nothing, Int, Int] = files.foldLeft(ZArrow.identity[Int]) {
     case (acc, item) => acc >>> ZArrow.lift(_ => worker(item._1))
   }
   val sumArr: IO[Nothing, Int] = arrWorkers.run(0)
```

Test Env:
Ubuntu 18.04 
Java 11: OpenJDK Runtime Environment (build 11.0.6+10-post-Ubuntu-1ubuntu118.04.1)
Scala 2.13.1

Bench setup:
workers = 50 
Factorial is random in range [8...12]

Runtime results (microseconds, smaller better):
Plain - 4197 us
ZIO - 7311 us
ZArrow - 6291 us

`ZArrow` wins over `ZIO` in this scenario for a large enough workloads !

### CODE REVIEW

File: project/BuildHelper.scala [hunk 1]
```diff
@@ -21,8 +21,8 @@ object BuildHelper {
     "-Ywarn-numeric-widen",
     "-Ywarn-value-discard",
     "-unchecked",
-    "-deprecation",
-    "-Xfatal-warnings"
+    "-deprecation"
+    // "-Xfatal-warnings"
```
Comment (adamgfraser) [inline project/BuildHelper.scala]: Any reason for getting rid of this? I find it is generally helpful in maintaining code quality.

File: bench/src/main/scala/BenchUtils.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
+package bench
+
+import java.util.concurrent.TimeUnit
+
+import zio.arrow._
+import FileUtils._
+
+object BenchUtils {
```
Comment (adamgfraser) [inline bench/src/main/scala/BenchUtils.scala]: Would suggest that we add a `jmh` dependency and use that for the benchmarks.

File: project/BuildHelper.scala (hunk 1 above)
Comment (tampler) [reply project/BuildHelper.scala]: Sure, I'll revert in the final version. That's just work-in-progress

File: bench/src/main/scala/BenchUtils.scala (hunk 2 above)
Comment (tampler) [reply bench/src/main/scala/BenchUtils.scala]: Yes, will do and will port existing bemchmarks too. Thanks for advice !

**Comment by tampler** [comment]
@adamgfraser pls review my changes 

How do I run the `jmh` benchmarks? I can compile, but can't run them now

**Comment by adamgfraser** [comment]
@tampler https://github.com/zio/zio/blob/master/benchmarks/README.md

**Comment by tampler** [comment]
@jdegoes  @adamgfraser Finished JMH benchmark. Results are below:
```bash
[info] # JMH version: 1.21
[info] # VM version: JDK 11.0.6, OpenJDK 64-Bit Server VM, 11.0.6+10-post-Ubuntu-1ubuntu119.10.1
[info] Benchmark                    Mode  Cnt         Score   Error  Units
[info] SocketBenchmark.arrowBench  thrpt       74218017.412          ops/s
[info] SocketBenchmark.plainBench  thrpt           2875.836          ops/s
[info] SocketBenchmark.zioBench    thrpt       62868643.228          ops/s
```

`ZArrow` is **18%** faster than `ZIO` even without any optimizations !
Congrats, gentlemen! Meet a new tech for high performance applications !

**Comment by tampler** [comment]
Updated benchmark results by adding `unsafeRun`:
```bash
[info] Benchmark                    Mode  Cnt      Score   Error  Units
[info] SocketBenchmark.arrowBench  thrpt       12724.911          ops/s
[info] SocketBenchmark.plainBench  thrpt       13462.444          ops/s
[info] SocketBenchmark.zioBench    thrpt        9855.602          ops/s
```

**Comment by adamgfraser** [comment]
Awesome! Is this ready to go or are you still adding more to it?

**Comment by tampler** [comment]
Yes, you may close it now. Adam, lets have a deal: If you put a `WIP` on the name - its a work-under-progress and you don't merge it. If  you don't see `WIP` - pls review thoroughly and merge it if its ok

Enough for this PR. My latest results are:

```bash
[info] Benchmark                    Mode  Cnt      Score   Error  Units
[info] SocketBenchmark.arrowBench  thrpt       52903.643          ops/s
[info] SocketBenchmark.plainBench  thrpt       63095.033          ops/s
[info] SocketBenchmark.zioBench    thrpt       45068.404          ops/s
```
And `ZArrow` is **17% faster** `ZIO`

**Comment by adamgfraser** [comment]
@tampler Deal!

**Pull request merged**